sycl/doc/extensions/proposed/sycl_ext_oneapi_cuda_cluster_group.asciidoc

```diff
@@ -0,0 +1,394 @@
+        using range_type = range<Dim>;
+        using linear_id_type = size_t;
+
+        linear_id_type get_group_linear_id() const;
+
+        linear_id_type get_local_linear_id() const;
+
+        range_type get_group_range() const;
+
+        id_type get_group_id() const;
+
+        id_type get_local_id() const;
+
+        range_type get_local_range() const;
+
+        linear_id_type get_local_linear_range() const;
+
+        linear_id_type get_group_linear_range() const;
+
+        bool leader() const;
+    }
+----
+
+
+[%header,cols="5,5"]
+|===
+|Method
+|Description
+
+|`linear_id get_group_linear_id() const`
+|Returns the linearized id of the calling work-group within the cluster
+
+|`linear_id get_local_linear_id() const`
+|Returns the linearized index of the calling work-item within the cluster
+
+|`range_type get_group_range() const`
+|Returns the number of work-groups in each dimension within the cluster
+
+|`id_type get_group_id() const`
+|Returns the id of the calling work-group along each dimension within the cluster
+
+|`id_type get_local_id() const`;
+|Returns the id of calling work-item along each dimension within the cluster
+
+|`range_type get_local_range() const`;
+|Returns the number of work-items along each dimension within the cluster.
+
+|`linear_id_type get_local_linear_range() const`;
+|Returns a linearized version of the `range_type` returned by  `get_local_range`
+
+|`linear_id_type get_group_linear_range() const`;
+|Returns a linearized version of the `range_type` returned by  `get_group_range`
+
+|`bool leader() const`;
+|Returns true for exactly one work-item in the cluster, if the calling work-item
+is the leader of the cluster group. The leader is guaranteed to be the work-item
+for which `get_local_linear_id` return 0.
+|===
+
+
+== Accessing another work-group's local memory
+
+Work-groups within the cluster have the ability to access another work-group's 
+local memory. Typically addresses which reside in the local memory of a 
+work-group can only be accessed by the work-items of that work-group. 
+Therefore, to access another work-group's local memory, the address needs to be
+mapped such that the address in another work-group is accessible within the 
+calling work-item. Further, to access another another work-group's local memory,
+all the work-groups within the cluster must exist and the work-groups should
+not cease to exist before all the memory operations are completed. This can be 
+ensured by synchronizing all the work-items within the cluster before and after
+the local memory operations, using `group_barrier`.
+
+A member function of the `cluster_group` class; 
+`map_cluster_local_pointer` will perform the mapping and return a pointer 
+which can then be dereferenced by the calling work-item. 
+
+
+[%header,cols="10,5"]
+|===
+|Method
+|Description
+
+|T*  map_cluster_local_pointer(T* addr, size_t group_id)
+|Accepts the equivalent address to the memory location relative to the calling 
```

Comment:
It doesn't say anything about address space. I assume this means general SYCL rules apply both in/out pointer are generic address space? And therefore this will map to `__nvvm_mapa`?
Cutlass calls `mapa.shared::cluster` (https://github.com/NVIDIA/cutlass/blob/56b46e2d13875b46b8f6a03f9f5ac91e2bfdc01a/include/cute/arch/cluster_sm90.hpp#L169). Would it be recommended to directly call `__nvvm_mapa_shared_cluster` instead of `map_cluster_local_pointer` for best performance. 
Why do we have this function at all and don't just let users directly call the built-in. Seems like this function is anyhow a very low level function.

Reply:
Hi Roland, thanks for your feedback.

You raise a valid point, and this  should be differentiating interfaces accepting generic address space and shared address space (read decorated), and accepting multi_ptr would be a good way to do that.

Yes, this function will practically be  just a wrapper around `__nvvm_mapa*`, as it's interface was designed in a bottom up fashion so to speak, by checking what PTX requires and thus what we need to pass from the top level.

But I would still advocate against relying (or rather suggesting ) on the users to inline asm or use compiler intrinsics, as it would not be a very SYCL way of doing things, and would add lead to complexity in the user's code.

Reply:
I agree it is nicer to have a proper SYCL interface. But for these kind of low level functions the advantage is small and therefore we should only do it if it doesn't come at a performance impact. In case you plan to add the version with shared address space in a later version, I would prefer if you wait to add the whole function in that later version. Because I think it is better to recommend to use users inline asm or intrinsics over incurring a performance impact.

Reply:
This raises a really good point, I've been discussing this with @Naghasan and we came up with some ideas for how to address this.

For the current interface, where we are simply using raw pointers, these would be in the generic address space, so this would map to a `mapa` instruction. We think this could be optimised to `mapa.shared::cluster` either by the LLVM NVPTX backend or perhaps in the PTX assembler, this would require some investigation, but without these optimizations we agree this would not be as efficient as calling the PTX instruction directly in asm. This would keep things simple, though it would be hiding the fact that a generic pointer may be to local memory and that may point to different local address spaces, which are CUDA-specific.

As @AD2605 mentioned we have considered having a `multi_ptr` interface as well and this may be a way to represent this. We could add an additional address space, though this would be a big change to introduce for this feature. The other thing to consider here is that LLVM currently only has one local memory address space so it cannot distinguish between `.shared::cta` and `.shared::cluster`, so again it would rely on the NVPTX backend to lower this correctly, perhaps inferring the correct address space through the explicit use of `mapa.shared::cluster`.

I think we tend to agree with @rolandschulz's conclusion that perhaps we are trying to expose a CUDA API which is too low level to abstract in SYCL without expanding the SYCL memory model. Another option we have considered, which we wanted to explore more in the future was whether it would be better to rather than exposing the low-level API for mapping local memory between work-groups in a cluster-group, we instead provide group algorithms for the `cluster_group` for all the existing algorithms. The mapping of local memory between work-groups in a cluster-group would then become an implementation detail of these algorithms. I think this would also address @Pennycook's concern above about having functions which appear like group functions but don't follow the same semantics, i.e. are not cooperative.

Our initial goal was to try to align with CUDA here, but perhaps in this case it's better to have a more high-level abstraction. We could still have a SYCLcompat interface which provides a more low-level mapping of the CUDA APIs, though there's always a concern there that this would miss out on some potential optimisation opportunities.

We would like to close this PR soon so we can have this in for the 2025.0 release, so I would propose that we keep the current interface, clarify that the pointers will be assumed to be in the generic address space, and add a known issue that we'd like to continue to explore the design space for this. Alternatively we could remove these APIs in the current version entirely if there are concerns with having it in it's current form.

Reply:
@rolandschulz ,  I have added this in the known issues section here - [824af1e](https://github.com/intel/llvm/pull/13594/commits/824af1e30203f42612a8c13bbe4740c91aeb1a2f)